now downgrading remark headers.

joelmartinez opened the pull request:
Per requirements here: 
https://ceapex.visualstudio.com/Engineering/_queries/edit/63993

There is a dependent PR in the templates that should be merged in concert with this change:
https://ceapex.visualstudio.com/Engineering/_git/docs-ui/pullrequest/8378?_a=overview 

Please take a look at the unit tests to see examples of scenarios that this change handles. But a few assumptions and notes:

- I'm making a blanket assumption that all header levels should be downgraded, so the original work item only mentions h2 and h3, but this works on all headers.
- Will only transform h1 through h5 … this is because [html only supports up to h6](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements), so an h5 is converted to h6, but h6 would remain unchanged.
- Line endings will be normalized to the `Environment.NewLine` for whatever platform runs _Ecma2Yaml_ on CI … my assumption here is that the [git repo line endings configuration](https://help.github.com/articles/dealing-with-line-endings/) will be set accordingly, so _Ecma2Yaml_ should respect the platform.
- The code will **only** downgrade markdown headers that are at **the start** of a line. So that means that headers in a code section (indented by 4 spaces), or in a blockquote (line starts with a `>`), or in a ordered or unordered list **will not** be transformed. 
- Additionally, the code makes a special provision for content in a [GFM-style fenced code block](https://help.github.com/articles/creating-and-highlighting-code-blocks/#fenced-code-blocks) … this content will not be transformed either, which allows for documentation of markdown itself, or CLI scripts where comments start with `#`.

Please let me know if any of these assumptions are incorrect, or if anything should otherwise change. Thanks!

joelmartinez committed:
now downgrading remark headers.
Closes https://ceapex.visualstudio.com/Engineering/_queries/edit/63993

TianqiZhang commented on ECMA2Yaml/ECMAHelper/ECMALoader.docs.cs:
```diff
@@ -121,6 +123,54 @@ public Docs LoadDocs(XElement dElement)
+        /// <summary>Modifies the array if a header of the given size is found</summary>
+        private static void ReplaceTriggered(string[] lines, int headerCount, ref bool replaceTriggered)
+        {
+            string headerPrefix = new string('#', headerCount);
```
it's better to use a const string array like {"", "#", "##",...} instead of creating a lot of string objects

joelmartinez replied:
I've changed this to use a static readonly array

TianqiZhang commented on ECMA2Yaml/ECMAHelper/ECMALoader.docs.cs:
```diff
@@ -121,6 +123,54 @@ public Docs LoadDocs(XElement dElement)
+                if (inCodeFence)
+                    continue;
+
+                if (line.StartsWith(headerPrefix) && line.Length > headerCount)
```
markdown engines like markdig often allows a few (less than 4) spaces in front of '#'s. Which means the following lines will all render to `<h2>`:
```
## header
 ## header with 1 space
  ## header with 2 spaces
   ## header with 3 spaces
```

joelmartinez replied:
The code now supports this, along with corresponding unit test

TianqiZhang commented on ECMA2Yaml/ECMAHelper/ECMALoader.docs.cs:
```diff
@@ -121,6 +123,54 @@ public Docs LoadDocs(XElement dElement)
+        /// <summary>Downgrades markdown headers from 1 - 5. So a `#` becomes `##`, but `######` (ie. h6) remains the same.</summary>
+        /// <param name="remarksText">A string of markdown content</param>
+        public static string DowngradeMarkdownHeaders(string remarksText)
+        {
```
the original template logic only triggers if `<h2>` is detected in remarks, otherwise everything remain untouched. We should follow this to keep compatibility.

joelmartinez replied:
The code will now only trigger if it finds an H2 in the content

joelmartinez committed:
markdown headers put in readonly string array.
To avoid excessive object allocations

joelmartinez committed:
header downgrade changes ...
- Now only triggers if an 'h2' is in the content
- Now supports up to three spaces in front of the '#'

TianqiZhang commented:
change merged, thanks a lot @joelmartinez !
